katharsis-core/src/main/java/io/katharsis/module/ModuleRegistry.java

```diff
@@ -390,65 +403,122 @@ public ResourceInformationBuilder getResourceInformationBuilder() {
+				if (repositoryInformation instanceof ResourceRepositoryInformation) {
+					resourceRepositoryInformation = (ResourceRepositoryInformation) repositoryInformation;
+					Object repository = repositoryImplementations.get(resourceRepositoryInformation);
+					resourceEntry = setupResourceRepository(resourceRepositoryInformation, repository);
+				}
+				else {
+					RelationshipRepositoryInformation relationshipRepositoryInformation = (RelationshipRepositoryInformation) repositoryInformation;
+					Object repository = repositoryImplementations.get(repositoryInformation);
+					setupRelationship(relationshipEntries, relationshipRepositoryInformation, repository);
+				}
 			}
-			if(repository instanceof ResourceRegistryAware){
-				((ResourceRegistryAware)repository).setResourceRegistry(resourceRegistry);
+
+			if (resourceRepositoryInformation == null) {
+				ResourceInformation resourceInformation = getResourceInformationBuilder().build(resourceClass);
+				resourceRepositoryInformation = new ResourceRepositoryInformationImpl(resourceClass,
+						resourceInformation.getResourceType(), resourceInformation);
 			}
+
+			RegistryEntry registryEntry = new RegistryEntry(resourceRepositoryInformation, resourceEntry, relationshipEntries);
+			resourceRegistry.addEntry(resourceClass, registryEntry);
 		}
+	}
 
-		for (Map.Entry<ResourceRepositoryInformation, Object>  entry: resourceRepositories.entrySet()) {
-			ResourceRepositoryInformation resourceRepositoryInfo = entry.getKey();
-			final Object repository = entry.getValue();
-			Class<?> resourceClass = resourceRepositoryInfo.getResourceInformation().getResourceClass();
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private ResourceEntry setupResourceRepository(ResourceRepositoryInformation resourceRepositoryInformation,
+			Object repository) {
+		final Object decoratedRepository = decorateRepository(repository);
+		RepositoryInstanceBuilder repositoryInstanceBuilder = new RepositoryInstanceBuilder(null, null) {
 
-			RepositoryInstanceBuilder repositoryInstanceBuilder = new RepositoryInstanceBuilder(null, null) {
+			@Override
+			public Object buildRepository() {
+				return decoratedRepository;
+			}
+		};
 
-				@Override
-				public Object buildRepository() {
-					return repository;
-				}
-			};
+		if (ClassUtils.getAnnotation(decoratedRepository.getClass(), JsonApiResourceRepository.class).isPresent()) {
+			return new AnnotatedResourceEntry(repositoryInstanceBuilder);
+		}
+		else {
+			return new DirectResponseResourceEntry(repositoryInstanceBuilder);
+		}
+	}
 
-			ResourceEntry resourceEntry;
-			if (ClassUtils.getAnnotation(repository.getClass(), JsonApiResourceRepository.class).isPresent()) {
-				resourceEntry = new AnnotatedResourceEntry(repositoryInstanceBuilder);
-			}
-			else {
-				resourceEntry = new DirectResponseResourceEntry(repositoryInstanceBuilder);
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private Object decorateRepository(Object repository) {
```

Comment:
This function is

1. A bit too long
2. The use of `instanceof` is troubling
3. These conditional blocks are nearly identical

Applying a common interface to both of these repository type is the way to go.